project/src/main/java/com/google/sps/servlets/BuyServlet.java

```diff
@@ -0,0 +1,128 @@
+/**
+ * servlet for buying a new investment
+ * returns an Investment object representing the just purchased investment
+ */
+@WebServlet("/buy")
+public class BuyServlet extends HttpServlet {
+
+  private static final Logger LOGGER = Logger.getLogger(BuyServlet.class.getName());
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        DataSource pool = (DataSource) request.getServletContext().getAttribute("db-connection-pool");
+
+        try (Connection conn = pool.getConnection()) {
+            long user = Long.parseLong(request.getParameter("user"));
+            long competition = Long.parseLong(request.getParameter("competition"));
+            String searchQuery = request.getParameter("search");
+            int amtInvested = Integer.parseInt(request.getParameter("amount"));
+            Investment investment = addInvestment(conn, user, competition, searchQuery, amtInvested);
+            if (investment == null) {
+                response.setStatus(500);
+                response.getWriter().write("Timeout while fetching investment. Try again later.");
+            } else {
+                Gson gson = new Gson();
+                response.setContentType("application/json");
+                response.getWriter().println(gson.toJson(investment));
+            }
+        } catch (SQLException ex) {
+            LOGGER.log(Level.WARNING, "Error while attempting to buy investment.", ex);
+            response.setStatus(500);
+            response.getWriter().write("Unable to successfully add investment");
+        }
+    }
+
+    private Investment addInvestment (
+        Connection conn, long user, long competition, String searchQuery, int amtInvested
+        ) throws SQLException, IOException {
+        
+        InvestmentCalculator calc = new InvestmentCalculator();
+        Long currentDateSeconds = calc.getLatestDate();
+        Date currentDate = new Date(currentDateSeconds * 1000);
+
+        // check if we already have data for this search
+        ImmutableList<Long> data = calc.getInvestmentDataIfExists(searchQuery);
+        if (data.isEmpty()) {
+            // if not, fetch using cloud function and wait for some set time
+            data = calc.createAndFetchInvestmentData(searchQuery, data);
+        }
+
+        if (data.isEmpty()) {
+            //we timed out. return null and then send response - DO NOT add to investments db
+            return null;
+        }
+
+        String stmt = String.format("INSERT INTO investments (user, competition, google_search, " 
+            + "invest_date, sell_date, amt_invested) VALUES (%d, %d, '%s', DATE '%tF', NULL, %d);",
+            user, competition, searchQuery, currentDate, amtInvested);
+
+        try (PreparedStatement investmentStmt = conn.prepareStatement(stmt);) {
+            // Execute the statement
+            investmentStmt.execute();
+        }
+        String getId = "SELECT LAST_INSERT_ID();";
```

Comment:
Think Tex mentioned there was a way to extract this information from the ResultSet (my exact recollection is not 100%)

Reply:
I thought so too, but I found Phoebe's code from the time he mentioned it and she did it this way so I guessed there was a reason for that. I can check with her.

Reply:
Phoebe thought this was what Tex said so that's all the info I have.

Reply:
You can do

```java
Statement stmt = db.prepareStatement(query, Statement.RETURN_GENERATED_KEYS);
stmt.execute();
ResultSet rs = stmt.getGeneratedKeys();
```

Reply:
Fixed

Reply:
nice